consensus/istanbul/validator.go

```diff
@@ -44,12 +70,19 @@ type Validator interface {
 
 	BLSPublicKey() blscrypto.SerializedPublicKey
 
+	BLSPublicKeyUncompressed() []byte
+
 	// Serialize returns binary reprenstation of the Validator
 	// can be use used to instantiate a validator with DeserializeValidator()
 	Serialize() ([]byte, error)
 
 	// AsData returns Validator representation as ValidatorData
 	AsData() *ValidatorData
+
+	// AsData returns Validator representation as ValidatorData
+	AsDataWithCache() *ValidatorDataWithCache
+
+	CacheUncompressed()
```

Comment:
tbh, this looks rather ugly. We have a data type `ValidatorSet` where some implementation of it `defaultSet` has a derived field that it caches. Which is an implementation detail, not a design one.

Do all implementations need to have a cache?

I understand this is convenient, for the purppose of snapshot caching this values.. but it's ugly for a encapsulation/separation of cocerns PoV.

Still, given we are rather on a hurry, and i don't have a better proposal for the moment, we can go ahead. but still.. i want to make that clear

Reply:
Perhaps we should change all precompiles that depend on snapshots to only work for a limited amount of epochs, (then we can keep them all in memory). Then we could remove `ValidatorDataWithCache`. To remove `CacheUncompressed` from here, we could just store them in the valset instead, or the `defaultSet` could contain a list of `defaultValidator`...